### DISCUSSION TITLE
[deliver] Call for testers - a brand new screenshots synchronisation logic

### DISCUSSION THREAD

**Discussion by ainame** [fastlane/fastlane#19204] [Show and tell]
Hi there!👋 This is @ainame a core contributor of fastlane. I'm pleased to announce that we are going to be introducing the new feature in `deliver` that makes it easier to keep screenshots up-to-date with App Store Connect 📷  To do so, we are calling for testers of it. It is written from scratch so that we expect some unknown issues may exist. I'd love to hear your feedback.

This is the PR that introduce new logic.
https://github.com/fastlane/fastlane/pull/19070

## Why new feature? 🆕  

`deliver` has been helping developers to submit iOS apps to App Store Connect and update metadata and screenshots to save their time since it's built. However, I see a difficulty in the use of screenshots update functionality when it comes to a large project that has so many screenshots.

This is the pseudocode of the current logic to upload screenshots. Unless you enable `overwrite_screenshots`, it's not going to delete old screenshots so that you'll see the mix of old and new screenshots uploaded on the App Store even if you want to replace old ones with new ones.

```ruby
if optons[:overwrite_screenshots]
  delete_all()
end

for local_screenshot in local_screenshots 
  next if screenshot_exists_on_app_store_connect?(local_screenshot)
  upload(local_screenshot) 
end 
```

In contrary, the new logic can recognise what's changed from the ones on App Store Connect within your screenshot directory meaning it gets sync your screenshot directory with App Store Connect. This way you can delete and upload only what really needs to do so. It's faster, more intuitive, and more efficient. Most importantly it can save the number of API calls (App Store Connect API has rate limit on API calls) when possible. 

```ruby
# Exlude screenshots from the collection of `Spaceship::ConnectAPI::AppScreenshots` objects, which don't exist on local
screenshots_to_delete = remote_screenshots - local_screenshots

for remote_screenshot in screenshots_to_upload
  delete(remote_screenshot) 
end

# Exclud screenshots from the collection of `Deliver::AppScreenshot` objects, which don't exist on App Store Connect 
screenshots_to_upload = local_screenshots - remote_screenshots

for local_screenshot in screenshots_to_upload
  upload(local_screenshot) 
end
```

## How to participate? ✅ 

This is a part of v2.191.0 so it's now released. To use the feature, you need to pass `true` to `sync_screenshots` option and explicitly declare the use of "beta" feature by feature toggle mechanism, which prevents public users from unintentional use of unstable feature. To do so, you also need to set truthy value to `FASTLANE_ENABLE_BETA_DELIVER_SYNC_SCREENSHOTS` environment variable. For your convenient, you can even set it in your Fastfile directly, like below.

```ruby
ENV["FASTLANE_ENABLE_BETA_DELIVER_SYNC_SCREENSHOTS"] = 'true'

upload_to_app_store(
  sync_screenshots: true
)
```

**Note that this feature aims to get your screenshots sync with App Store Connect so images only exist on App Store Connect may be lost by running `deliver` with `sync_screnshots` option.** To avoid that, please back up important images by yourself. (You can use `fastlane deliver download_screenshots` 🚀 )

## Give us feedback!👂 

I hope you like this new feature and I'd love to hear your feedback so that we can improve the feature. Please share your comments if you have issues and questions.

**Comment by ainame** [comment]
fastlane v2.191.0 was released🎉 You can try this out by `gem install fastlane` or `bundle update fastlane` now 🚀 
https://github.com/fastlane/fastlane/releases/tag/2.191.0

**Reply by vincentisambart** [reply]
> ```ruby
> ENV["FASTLANE_ENABLE_BETA_DELIVER_SYNC_SCREENSHOTS"] = true
> ```

That causes a "no implicit conversion of true into String" error so I guess that should be `ENV["FASTLANE_ENABLE_BETA_DELIVER_SYNC_SCREENSHOTS"] = '1'` (or more precisely any string not any of "no", "false", "off", "0").

**Reply by ainame** [reply]
Thank you for that point🙂 I really didn't test that part with inline code using `ENV[]`😓

**Comment by LinusU** [comment]
I just tested this out on ~70 apps, and it seems to work great! Thank you for adding this feature 🙏  🎉

In fact, I was very surprised to learn that this isn't the default way which resulted in us submitting (and getting approved 😂) an app where every screenshot was duplicated. This has been in a release for around a year now, are there any plans to remove the requirement of a `FASTLANE_ENABLE_BETA_DELIVER_SYNC_SCREENSHOTS` flag?

Are there any plans on making this the default without having to pass `--sync_screnshots`? It seems like it's quite surprising when it doesn't work like this, no?

**Comment by squarefrog** [comment]
How does `sync_screenshots` determine if the images have already been uploaded or not? We tried this out in the hopes it would fix an issue we have where `deliver` uploads all the screenshots successfully, but then times out after an hour as 40% of the screenshots get stuck in processing forever.

When we tried this last night, I first uploaded all 128 of our screenshots, then ran the following in a test lane:
```
  desc "Test deliver"
  lane :test_deliver do 
    ENV["FASTLANE_ENABLE_BETA_DELIVER_SYNC_SCREENSHOTS"] = 'true'

    deliver(
      precheck_include_in_app_purchases: false,
      app_version: "4.14.0",
      build_number: "161",
      skip_binary_upload: true,
      submit_for_review: false,
      reject_if_possible: true,
      automatic_release: false,
      phased_release: true,
      force: true,
      sync_screenshots: true,
      submission_information: {
        add_id_info_limits_tracking: true,
        add_id_info_serves_ads: false,
        add_id_info_tracks_action: true,
        add_id_info_tracks_install: true,
        add_id_info_uses_idfa: true
      }
    )
  end
```

The first thing it did was delete all our screenshots from App Store connect, then attempt to upload them from scratch, which caused the processing issue again 😭

**Reply by kuhnroyal** [reply]
Same here, decided to try this feature because of https://github.com/fastlane/fastlane/issues/20619 but it deleted everything and re-uploaded all screenshots.

It failed with
```
fastlane-2.209.1/spaceship/lib/spaceship/connect_api/api_client.rb:204:in `handle_response': The specified resource does not exist - There is no resource of type 'appScreenshots' with id 'a4f2a843-3575-476d-bf2a-ed23a2020a6f' (Spaceship::UnexpectedResponse)
```

And now on the next attempt, it is stuck in processing again after again deleting all screenshots.

**Comment by benlaverriere** [comment]
Is there testing or feedback that would help move this feature out of beta status? My team has had some trouble with the old/default behavior, so we've been looking at this one. If we can provide anything to help "promote" the feature, we'd be happy to help.

**Comment by Patricio879** [comment]
Thank you for being so generous with your knowledge. It's truly appreciated.
